crates/bevy_ecs/src/system/builder.rs

```diff
@@ -203,12 +389,67 @@ mod tests {
         world.spawn(A);
         world.spawn_empty();
 
-        let system = SystemBuilder::<()>::new(&mut world)
-            .local::<u64>()
-            .param::<Local<u64>>()
-            .build(multi_param_system);
+        let system = (LocalBuilder(0), ParamBuilder)
+            .build_state(&mut world)
+            .build_system(multi_param_system);
 
         let result = world.run_system_once(system);
         assert_eq!(result, 1);
     }
+
+    #[test]
+    fn param_set_builder() {
+        let mut world = World::new();
+
+        world.spawn((A, B, C));
+        world.spawn((A, B));
+        world.spawn((A, C));
+        world.spawn((A, C));
+        world.spawn_empty();
+
+        let system = (ParamSetBuilder((
+            QueryParamBuilder::new(|builder| {
+                builder.with::<B>();
+            }),
+            QueryParamBuilder::new(|builder| {
+                builder.with::<C>();
+            }),
+        )),)
+            .build_state(&mut world)
+            .build_system(|mut params: ParamSet<(Query<&mut A>, Query<&mut A>)>| {
+                params.p0().iter().count() + params.p1().iter().count()
+            });
+
+        let result = world.run_system_once(system);
+        assert_eq!(result, 5);
+    }
+
+    #[test]
+    fn param_set_vec_builder() {
```

Comment:
Question: what's the point of being able to extract a vec of queries? This seems...generally unhelpful.

Reply:
I plan to use it for scripting.  The idea is to load a script at runtime that does any number of queries, and then build a script runner system with a `Vec<Query<FilteredEntityMut>>` based on those queries (or `ParamSet<Vec<Query<FilteredEntityMut>>>` if they might conflict).  The script runner system would bind the queries in the script by indexing the `Vec`.  

(Buildable `Vec` and `ParamSet<Vec>` are the features I actually wanted here, and it would have been straightforward to implement `BuildableSystemParam` for them.  It just felt wrong to implement buildable `ParamSet<Vec>` without buildable `ParmSet<(tuple,)>`, and I didn't see a clean way to build tuples with a builder type.)  

I'm happy to split it out to a follow-up PR if that makes this easier to review!

Reply:
Yes please, I'd much prefer reviewing it as a separate PR :)